source/server/admin/stats_request.cc

```diff
@@ -0,0 +1,148 @@
+#include "source/server/admin/stats_request.h"
+
+namespace Envoy {
+namespace Server {
+
+StatsRequest::StatsRequest(Stats::Store& stats, bool used_only, bool json,
+                           Utility::HistogramBucketsMode histogram_buckets_mode,
+                           absl::optional<std::regex> regex)
+    : used_only_(used_only), json_(json), histogram_buckets_mode_(histogram_buckets_mode),
+      regex_(regex), stats_(stats) {}
+
+Http::Code StatsRequest::start(Http::ResponseHeaderMap& response_headers) {
+  if (json_) {
+    render_ =
+        std::make_unique<StatsJsonRender>(response_headers, response_, histogram_buckets_mode_);
+  } else {
+    render_ = std::make_unique<StatsTextRender>(histogram_buckets_mode_);
+  }
+
+  // Populate the top-level scopes and the stats underneath any scopes with an empty name.
+  // We will have to de-dup, but we can do that after sorting.
+  //
+  // First capture all the scopes and hold onto them with a SharedPtr so they
+  // can't be deleted after the initial iteration.
+  stats_.forEachScope(
+      [this](size_t s) { scopes_.reserve(s); },
+      [this](const Stats::Scope& scope) { scopes_.emplace_back(scope.getConstShared()); });
+
+  startPhase();
+  return Http::Code::OK;
+}
+
+bool StatsRequest::nextChunk(Buffer::Instance& response) {
+  if (response_.length() > 0) {
+    ASSERT(response.length() == 0);
+    response.move(response_);
+    ASSERT(response_.length() == 0);
+  }
+  while (response.length() < chunk_size_) {
+    while (stat_map_.empty()) {
+      switch (phase_) {
+      case Phase::TextReadouts:
+        phase_ = Phase::CountersAndGauges;
+        startPhase();
+        break;
+      case Phase::CountersAndGauges:
+        phase_ = Phase::Histograms;
+        startPhase();
+        break;
+      case Phase::Histograms:
+        render_->finalize(response);
+        return false;
+      }
+    }
+
+    auto iter = stat_map_.begin();
+    StatOrScopes variant = std::move(iter->second);
+    StatOrScopesIndex index = static_cast<StatOrScopesIndex>(variant.index());
+    switch (index) {
+    case StatOrScopesIndex::Scopes:
```

Comment:
In the future, is it possible to simply iterate through each scope? That should make the code more readable and simpler.

Reply:
possibly. We could certainly blend text-readouts and gauges/histograms, though I think it might be nicer to keep histograms separate. They are buffered in JSON anyway, but not in text output.